piptools/build.py

```diff
@@ -193,7 +243,13 @@ def _create_project_builder(
         yield build.ProjectBuilder(src_dir, runner=runner)
         return
 
-    with build.env.DefaultIsolatedEnv() as env:
+    maybe_pip_constrained_context = (
+        contextlib.nullcontext()
+        if upgrade_packages is None
+        else _temporary_constraints_file_set_for_pip(upgrade_packages)
+    )
+
+    with maybe_pip_constrained_context, build.env.DefaultIsolatedEnv() as env:
         builder = build.ProjectBuilder.from_isolated_env(env, src_dir, runner)
         env.install(builder.build_system_requires)
```

Comment:
So, while debugging via tmate, I discovered that on Windows, this raises a `SubprocessError` from somewhere within `pypa/build`. I'm not yet sure what's happening. The command was `pip install --pep517`something.. I didn't get to save it since the SSH session expires in about 40 minutes and I just lost it. I think, it's https://github.com/pypa/build/blob/88ae833/src/build/env.py#L257-L265.

This then causes the visible error we saw in the logs — `[RuntimeError("generator didn't yield")]`(https://github.com/python/cpython/blob/6bc3e83/Lib/contextlib.py#L143). I'm not entirely sure why, though — might be a CPython bug.

Now that I'm thinking about it, I'm curious if the order of the CMs in the `with` statement matters. Perhaps, `build.env.DefaultIsolatedEnv()` gets interrupted and processes the `__exit__()` first, suppressing the error. Then, the control flow of `_create_project_builder()` is interrupted, never reaching the `yield`, tripping the validation in the outer `@contextlib.contextmanager()` decorator.
Though, I don't understand why `SubprocessError` doesn't escape the `with` block… can it be https://github.com/python/cpython/blob/6bc3e83/Lib/contextlib.py#L163-L167?

Reply:
Calling `_create_project_builder()` or `build_project_metadata()` doesn't seem to be enough to reproduce this. I suspect, Click might be the cause.